tools/djxl_ng_main.cc

```diff
@@ -233,6 +235,107 @@ int DecompressJxlReconstructJPEG(const std::vector<uint8_t>& compressed,
   return EXIT_SUCCESS;
 }
 
+int DecompressJxlToPackedPixelFile(const std::vector<uint8_t>& compressed,
+                                   jxl::extras::PackedPixelFile& pixels,
+                                   JxlPixelFormat& format, JxlDecoderPtr dec,
+                                   JxlThreadParallelRunnerPtr runner) {
+  if (JXL_DEC_SUCCESS != JxlDecoderSetParallelRunner(dec.get(),
+                                                     JxlThreadParallelRunner,
+                                                     runner.get())) {
+    fprintf(stderr, "JxlEncoderSetParallelRunner failed\n");
+    return EXIT_FAILURE;
+  }
+  if (JXL_DEC_SUCCESS !=
+      JxlDecoderSubscribeEvents(dec.get(),
+                                JXL_DEC_BASIC_INFO | JXL_DEC_COLOR_ENCODING |
+                                    JXL_DEC_FRAME | JXL_DEC_FULL_IMAGE)) {
+    fprintf(stderr, "JxlDecoderSubscribeEvents failed\n");
+    return EXIT_FAILURE;
+  }
+
+  // Reading compressed JPEG XL input and decoding to pixels
+  if (JXL_DEC_SUCCESS !=
+      JxlDecoderSetInput(dec.get(), compressed.data(), compressed.size())) {
+    fprintf(stderr, "Decoder failed to set input\n");
+    return EXIT_FAILURE;
+  }
+  // TODO(firsching): handle boxes as well (exif, iptc, jumbf and xmp).
+  for (;;) {
+    JxlDecoderStatus status = JxlDecoderProcessInput(dec.get());
+    if (status == JXL_DEC_ERROR) {
+      fprintf(stderr, "Failed to decode image\n");
+      return EXIT_FAILURE;
+    } else if (status == JXL_DEC_NEED_MORE_INPUT) {
+      fprintf(stderr, "Error, already provided all input\n");
+      return EXIT_FAILURE;
+    } else if (status == JXL_DEC_BASIC_INFO) {
+      if (JXL_DEC_SUCCESS != JxlDecoderGetBasicInfo(dec.get(), &pixels.info)) {
+        fprintf(stderr, "JxlDecoderGetBasicInfo failed\n");
+        break;
```

Comment:
What's the principle that decides that we should break here rather than returning EXIT_FAILURE?